revocation/result/results.go

```diff
@@ -79,6 +87,70 @@ func NewServerResult(result Result, server string, err error) *ServerResult {
 	}
 }
 
+// CRLReasonCode is CRL reason code (See RFC 5280, section 5.3.1)
+type CRLReasonCode int
+
+const (
+	CRLReasonCodeUnspecified          CRLReasonCode = 0
+	CRLReasonCodeKeyCompromise        CRLReasonCode = 1
+	CRLReasonCodeCACompromise         CRLReasonCode = 2
+	CRLReasonCodeAffiliationChanged   CRLReasonCode = 3
+	CRLReasonCodeSuperseded           CRLReasonCode = 4
+	CRLReasonCodeCessationOfOperation CRLReasonCode = 5
+	CRLReasonCodeCertificateHold      CRLReasonCode = 6
+	CRLReasonCodeRemoveFromCRL        CRLReasonCode = 8
+	CRLReasonCodePrivilegeWithdrawn   CRLReasonCode = 9
+	CRLReasonCodeAACompromise         CRLReasonCode = 10
+)
+
+// String provides a conversion from a ReasonCode to a string
+func (r CRLReasonCode) String() string {
+	switch r {
+	case CRLReasonCodeUnspecified:
+		return "Unspecified"
+	case CRLReasonCodeKeyCompromise:
+		return "KeyCompromise"
+	case CRLReasonCodeCACompromise:
+		return "CACompromise"
+	case CRLReasonCodeAffiliationChanged:
+		return "AffiliationChanged"
+	case CRLReasonCodeSuperseded:
+		return "Superseded"
+	case CRLReasonCodeCessationOfOperation:
+		return "CessationOfOperation"
+	case CRLReasonCodeCertificateHold:
+		return "CertificateHold"
+	case CRLReasonCodeRemoveFromCRL:
+		return "RemoveFromCRL"
+	case CRLReasonCodePrivilegeWithdrawn:
+		return "PrivilegeWithdrawn"
+	case CRLReasonCodeAACompromise:
+		return "AACompromise"
+	default:
+		return fmt.Sprintf("invalid reason code with value: %d", r)
+	}
+}
+
+// CRLResult encapsulates the result of a CRL check
+type CRLResult struct {
```

Comment:
```go

type ServerResult struct {
	// Result of revocation for this server (Unknown if there is an error which
	// prevents the retrieval of a valid status)
	Result Result

	// Server is the URI associated with this result. If no server is associated
	// with the result (e.g. it is a root certificate or no OCSPServers are
	// specified), then this will be an empty string ("")
	Server string

	// Error is set if there is an error associated with the revocation check
	// to this server
	Error error
}

// CRLResult encapsulates the result of a CRL check
type CRLResult struct {
	// Result of revocation for this URI
	Result Result

	// ReasonCode is the reason code for the CRL status
	//
	// The reason code is only set if the certificate was revoked
	ReasonCode CRLReasonCode

	// RevocationTime is the time at which the certificate was revoked
	RevocationTime time.Time

	// URI is the URI to download the CRL
	URI string

	// Error is set if there is an error associated with the revocation check
	Error error
}
```
If we combine them, it will be confusing because `ReasonCode` is only for CRL, and `Server` is only for OCSP. We may need another field, such as `RevocationType`, to distinguish between OCSP and CRL. Also `ServerResult` is not clear name for CRL as CRL is a static file and the result is an analysis from the CRL file. Could you provide a clearer suggestion?

Reply:
Both CRL and [OCSP](https://datatracker.ietf.org/doc/html/rfc6960#section-4.2.1) have `ReasonCode`. I think Server is basically an result from one URI (out of many) specified for OCSP checking.

Yes I do see the need of

Reply:
But the definition of `ReasonCode` for CRL and OCSP are different, so we cannot share the same field for CRL and OCSP?

Reply:
For example:
```go
type ServerResult struct {
	// Result of revocation for this server (Unknown if there is an error which
	// prevents the retrieval of a valid status)
	Result Result

	// Server is the URI associated with this result. If no server is associated
	// with the result (e.g. it is a root certificate or no OCSPServers are
	// specified), then this will be an empty string ("")
	Server string

	// Error is set if there is an error associated with the revocation check
	// to this server
	Error error

        RevocationMethod Method

	// ReasonCode is the reason code for the CRL status
	//
	// The reason code is only set if the certificate was revoked
	ReasonCode CRLReasonCode
}
```

When `RevocationMethod` == OCSP, the user cannot use `ReasonCode`. Is it what you expected?

Reply:
Is it difficult for the caller to decide which field to use?

Reply:
As per [rfc6960#section-4.2.1](https://datatracker.ietf.org/doc/html/rfc6960#section-4.2.1) `revocationReason` is `CRLReason`. So ReasonCode can be used for both OCSP and CRL. Also, the values for is described in [rfc5280#section-5.3.1](https://datatracker.ietf.org/doc/html/rfc5280#section-5.3.1).

Reply:
Merged `ServerResult` with `CRLResult`. Also removed the `ReasonCode` and `RevocationTime` as we may don't need them now.